# Guided Project: Exploring Ebay Car Sales Data

The goal of this project was to clean the data and conduct a thorough analysis of the used car listings from the data set eBay Kleinanzeigen. In order to conduct this analysis, this project aims to utilize pandas in order to work with the data more efficiently. 

## Step 1: Import Necessary Libraries and Files

In [1]:
import pandas as pd 
import numpy as np 
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')


After importing the necessary libraries and reading in the data, I wanted to observe what the first 5 rows of the auto data looked like as well as the information of the data. We see that the object is dataframe class with 20 columns. Also, the columns consist of object (strings) and int64 objects. We also see that in order to work with the data, we will need to convert most of the object types into numeric types (in this case float) and further clean the strings in order to conduct calculations and analysis. Finally, we see that the colum names use camelcase and we will convert to snakecase, as well as there being null values. 

In [2]:
autos
print(autos.head())
print(autos.info())

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

In [3]:
autos.columns
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.columns
autos.head()


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In the above cell I changed the column names from the form camelcase to snakecase. This was done in order to be more descriptive.

## Step 2: Data Exploration 

The next step is to determine what other cleaning tasks that need to be done. In this step I looked for columns where all or almost all values were the same because it can be removed most of the time. Also, numeric data stored as text can also be cleaned and converted. Let's take a look at the data with the describe method for autos. 

In [4]:
autos.describe(include = 'all')


date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

For the autos dataframe, we can see that the columns have a lot of NaN values across the board. One key important note is that the seller, offer_type and num_photos have nearly all the same values.  Also, the price and odometer columns are numeric values stored as text. So, we will remove any non-numeric characters and convert the column to a numeric dtype. Let's start by first removing the unecessary columns in our analysis. Then, I will fix the price column and then the odometer column. 

In [5]:
autos['num_photos'].value_counts()
autos['offer_type'].value_counts()
autos['seller'].value_counts()
autos = autos.drop(['num_photos', 'offer_type', 'seller'], axis = 1)

### Part A) Fix Price Column 

In [6]:
autos['price'].head()
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(float)
autos['price'].head()

0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
Name: price, dtype: float64

### Part B) Fix Odometer Column 

In [7]:
autos['odometer'].head()
autos.rename({'odometer':'odometer_km'}, axis = 1 , inplace = True)
autos.columns


Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

### Part Bi) Odometer Column Analysis

In [8]:
autos['odometer_km'].head()
autos['odometer_km'] = autos['odometer_km'].str.replace(',', '').str.replace('km','').astype(float)
autos['odometer_km'].head()

0    150000.0
1    150000.0
2     70000.0
3     70000.0
4    150000.0
Name: odometer_km, dtype: float64

In [9]:
autos['odometer_km'].unique().shape
print('\n')
print(autos['odometer_km'].describe())
print('\n')
print(autos['odometer_km'].value_counts().head().sort_index(ascending = False))




count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
Name: odometer_km, dtype: int64


### Ai) Price Analysis

In [10]:
print(autos['price'].unique().shape)
print('\n')
print(autos['price'].describe())
print('\n')
print(autos['price'].value_counts().head(10))
print('\n')
print(autos['price'].value_counts().sort_index().head(10))
print('\n')
autos = (autos[autos['price'].between(1,351000)])
print('\n')
print(autos['price'].describe())



(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0.0       1421
500.0      781
1500.0     734
2500.0     643
1200.0     639
1000.0     639
600.0      531
800.0      498
3500.0     498
2000.0     460
Name: price, dtype: int64


0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
Name: price, dtype: int64




count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64


### Part C) Analysis of Price and Odometer Columns 

From the 'odometer_km' the values in the field are rounded. This means, most likely, there were pre-set options. One can also see that there are significant more high mileage and low mileage vehicles. There are 2,357 unique values in the column and also there are 1,421 cars listed with $0 price. Since this only makes roughly about 2 percent of the cars, we can remove these rows. 


In [11]:
print(autos[['date_crawled', 'ad_created', 'last_seen']][0:5])
print('\n')
print(autos['date_crawled'].str[0:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True))
print('\n')
print(autos['date_crawled'].str[0:10].value_counts(normalize = True, dropna = False).sort_values())
print('\n')
print(autos['ad_created'].str[0:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True))
print('\n')
print(autos['ad_created'].str[0:10].value_counts(normalize = True, dropna = False).sort_values())
print('\n')
print(autos['last_seen'].str[0:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True))
print('\n')
print(autos['last_seen'].str[0:10].value_counts(normalize = True, dropna = False).sort_values())
print('\n')
print(autos['registration_year'].describe())

          date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50


2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31

From our data, the site seemed to be crawled daily in March and April. Also, the distribution of listings each day seems to be roughly uniform. Also, from the registration year column, it seems that there are odd outliers like the minimum and maximum which suggest the first car was registered in the year 1000 and the latest car was registered in the year 9999.


In [12]:
print(autos['registration_year'].describe())
print('\n')
print(autos['registration_year'].value_counts())
IQR = 2008 - 1999
Outlier_Q1 = 1999 - 1.5 * IQR 
Outlier_Q3 = 2008 + 1.5 * IQR
print(Outlier_Q1, Outlier_Q3)

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


2000    3156
2005    2936
1999    2897
2004    2703
2003    2699
2006    2670
2001    2636
2002    2486
1998    2363
2007    2277
2008    2215
2009    2085
1997    1951
2011    1623
2010    1589
2017    1392
1996    1373
2012    1310
1995    1227
2016    1220
2013     803
2014     663
1994     629
2018     470
1993     425
2015     392
1992     370
1990     347
1991     339
1989     174
        ... 
1950       3
9999       3
1800       2
1951       2
1941       2
1934       2
1954       2
1957       2
1955       2
2019       2
1001       1
1939       1
4500       1
1953       1
1111       1
4800       1
5911       1
1943       1
1938       1
1929       1
2800       1
6200       1
4100       1
8888       1
1927       1
9000       1
1948       1
1000       1
1931       1
19

The lowest value I decided to use for comparison was 1986 and 2018 for the highest value because I calculated what the two outlier boundaries were. Anything lower than 1985 is an outlier and anything over 2021 is an outlier as well. Therefore, I will use the interval of comparison from (1986, 2018).

In [13]:
(~autos['registration_year'].between(1986,2018)).sum() / autos.shape[0]
autos = autos[autos['registration_year'].between(1986,2018)]
print(autos['registration_year'].value_counts(normalize = True))


2000    0.066165
2005    0.061553
1999    0.060735
2004    0.056668
2003    0.056584
2006    0.055976
2001    0.055263
2002    0.052118
1998    0.049540
2007    0.047737
2008    0.046437
2009    0.043712
1997    0.040902
2011    0.034026
2010    0.033313
2017    0.029183
1996    0.028785
2012    0.027464
1995    0.025724
2016    0.025577
2013    0.016835
2014    0.013900
1994    0.013187
2018    0.009853
1993    0.008910
2015    0.008218
1992    0.007757
1990    0.007275
1991    0.007107
1989    0.003648
1988    0.002830
1987    0.001509
1986    0.001509
Name: registration_year, dtype: float64


In [14]:
top_brands = autos['brand'].value_counts(normalize = True)
top_5_brands = top_brands[top_brands > .05].index
print(top_5_brands)
final_dictionary = {}
for b in top_5_brands: 
    brands = autos[autos['brand'] == b]
    mean_price = brands['price'].mean()
    final_dictionary[b] = float(mean_price)
final_dictionary

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


{'audi': 9228.046679499519,
 'bmw': 8260.968845565749,
 'ford': 3429.3833383640203,
 'mercedes_benz': 8394.748829953198,
 'opel': 2919.9413574312634,
 'volkswagen': 5295.595872235872}

From here, we can see that Audi, BMW and Mercedes have a higher average price than the other car brands listed. Afterwards, there is a distinct price gap between the other car brands. Opel and Ford are the cheapest while Volkswagen is in the middle. 

In [15]:
mean_mileage = {}
for b in top_5_brands: 
    brands = autos[autos['brand'] == b]
    mean_mile = brands['odometer_km'].mean()
    mean_mileage[b] = float(mean_mile)
mean_mileage
price = pd.Series(final_dictionary).sort_values(ascending = False)
mile = pd.Series(mean_mileage).sort_values(ascending = False)
brand_information = pd.DataFrame(mile, columns = ['Average_Km'])
brand_information['mean_price'] = price
print(brand_information)


                  Average_Km   mean_price
bmw            132845.948012  8260.968846
mercedes_benz  131162.246490  8394.748830
opel           129954.816381  2919.941357
audi           129480.269490  9228.046679
volkswagen     129349.385749  5295.595872
ford           125321.460912  3429.383338


Overall, we can see that BMW and Mercedes have the highest average km but are not necessarily the highest in average price. 